Functor documentation and a conversion utility (#24494)
idaholab/moose: framework/doc/content/syntax/FunctorMaterials/index.md

# FunctorMaterials

> 
> Functor material properties are properties that are evaluated on-the-fly.
> Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
> for the creation of objects with considerable inter-operability between systems.
> 
> ## Using functor materials
> 
> Much like regular materials declare regular material properties, functor materials declare functor material properties.

Author: ```suggestion

Much like regular materials declare regular material properties, functor materials declare functor material properties.
Functor material properties are properties that are evaluated on-the-fly.
Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
for the creation of objects with considerable inter-operability between systems.

## Using functor materials

```

> Much like regular materials declare regular material properties, functor materials declare functor material properties.

Author: done

Following a producer/consumer model, `FunctorMaterials` produce functor material properties, which kernels, boundary conditions, other
functor materials, etc, may consume.
Both regular and functor material properties support a wide array of data types (scalar, vector, tensors,
[automatic differentiation](automatic_differentiation/index.md) versions of all of these). Functor materials, because they are evaluated on-the-fly
and call other functors on-the-fly as well, do not require dependency ordering.

All functor materials support caching in some capacity. This can avoid expensive material property computations, but is
disabled by default due to the potential memory cost. The reader is referred to the [Functors caching documentation](syntax/Functors/index.md#caching)

!alert warning

> Functor materials are generally **NOT** compatible with regular materials, as regular material properties **CANNOT** be used

Author: reworded

in lieu of functor material properties. Functor material properties can only be used as regular material properties using a
[MaterialFunctorConverter.md] to process them.

Functor materials can be created within the `[FunctorMaterials]` block, and currently the `[Materials]` block as well,
but that syntax will be deprecated.

!alert note
If a Functor is reported as missing by the simulation, and it was supposed to be created by a `FunctorMaterial`,

> you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were

Author: ```suggestion
you may use the [!param](/Debug/SetupDebugAction/show_functors) parameter to get more information about what functors were
```

Reviewer: What I meant was that there's no indication that this is a parameter that needs to be in a `Debug` block.

> you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were

Author: added parameter after

> you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were

Reviewer: I put this comment on some outdated suggestion I think: the problem is more that there's no indication that this "show_functors" parameter is a `Debug` block parameter.

> you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were

Author: will change it

"geometric quantity" is based around an element, e.g. for an `FVElementalKernel`, the
value of a functor material property in a cell-averaged sense can be obtained by
the syntax

- `_foo(makeElemArg(_current_elem), determineState())`

where here `_foo` is a functor data member of the kernel, `makeElemArg` is a helper routine for creating a
functor element-based spatial argument, and `determineState()` is a helper routine for determining the correct
time state to evaluate at, e.g. the current time for an implicit kernel and the old time for an explicit kernel.
The functor material property system introduces APIs slightly different from the traditional
material property system for declaring/adding and getting properties. To add a
functor property:

- `addFunctorProperty<TYPE>`

where `TYPE` can be anything such as `Real, ADReal, RealVectorValue, ADRealVectorValue`
etc. To get a functor material property:

- `getFunctor<TYPE>`

It's worth noting that whereas the traditional regular material property system
has different methods to declare/get non-AD and AD properties, the new functor
system has single APIs for both non-AD and AD property types.

> Currently, functor material property evaluations are defined using the API:

Reviewer 2: It hasn't changed for a while (although it did change once before), so probably not

> Currently, functor material property evaluations are defined using the API:

Author: ```suggestion
Functor material property evaluations are defined using the API:
```

> Currently, functor material property evaluations are defined using the API:

Author: Removed current

# FunctorMaterials

Functor material properties are properties that are evaluated on-the-fly.

> Along with numerous classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows

Reviewer 2: ```suggestion
Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
```
Only three main systems are functors at the current time, so numerous feels like a bit too much. Maybe not forever though!

for the creation of objects with considerable inter-operability between systems.

## Using functor materials

Much like regular materials declare regular material properties, functor materials declare functor material properties.
Following a producer/consumer model, `FunctorMaterials` produce functor material properties, which kernels, boundary conditions, other
functor materials, etc, may consume.
Both regular and functor material properties support a wide array of data types (scalar, vector, tensors,
[automatic differentiation](automatic_differentiation/index.md) versions of all of these). Functor materials, because they are evaluated on-the-fly

> and call other functors on-the-fly as well, do not require dependency ordering.

Reviewer 2: Yes, another very important benefit which I often forget to bring up. Thanks for highlighting this!

All functor materials support caching in some capacity. This can avoid expensive material property computations, but is
disabled by default due to the potential memory cost. The reader is referred to the [Functors caching documentation](syntax/Functors/index.md#caching)

!alert warning
Functor materials are generally **NOT** compatible with regular materials, as regular material properties **CANNOT** be used
in lieu of functor material properties. Functor material properties can only be used as regular material properties using a
[MaterialFunctorConverter.md] to process them.

Functor materials can be created within the `[FunctorMaterials]` block, and currently the `[Materials]` block as well,

> but that syntax is considered for deprecation.

Reviewer 2: ```suggestion
but that syntax will be deprecated.
```

!alert note
If a Functor is reported as missing by the simulation, and it was supposed to be created by a `FunctorMaterial`,
you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were
created and requested.

## Developing with functor materials

Functor material properties are properties that are evaluated
on-the-fly. E.g. they can be viewed as functions of the current location in
space (and time). Functor material properties provide several overloads of the
`operator()` method for different "geometric quantities". One example of a

> "geometric quantity" is a `const Elem *`, e.g. for an `FVElementalKernel`, the

Reviewer 2: ```suggestion
"geometric quantity" is based around an element, e.g. for an `FVElementalKernel`, the
```

value of a functor material property in a cell-averaged sense can be obtained by
the syntax

> - `_foo(_current_elem)`

Reviewer 2: ```suggestion
- `_foo(makeElemArg(_current_elem), determineState())`
```

> where here `_foo` is a functor material property data member of the kernel. The

Reviewer 2: ```suggestion
where here `_foo` is a functor data member of the kernel, `makeElemArg` is a helper routine for creating a functor element-based spatial argument, and `determineState()` is a helper routine for determining the correct time state to evaluate at, e.g. the current time for an implicit kernel and the old time for an explicit kernel. The
```

functor material property system introduces APIs very similar to the traditional
material property system for declaring and getting properties. To declare a
functor property:

- `declareFunctorProperty<TYPE>`

where `TYPE` can be anything such as `Real, ADReal, RealVectorValue, ADRealVectorValue`
etc. To get a functor material property:

- `getFunctor<TYPE>`

It's worth noting that whereas the traditional regular material property system
has different methods to declare/get non-AD and AD properties, the new functor
system has single APIs for both non-AD and AD property types.

Currently, functor material property evaluations are defined using the API:

```c++

> template <typename T>
> template <typename PolymorphicLambda>
> void FunctorMaterialProperty<T>::
> setFunctor(const MooseMesh & mesh,
>            const std::set<SubdomainID> & block_ids,
>            PolymorphicLambda my_lammy);

Reviewer 2: ```suggestion
template <typename T, typename PolymorphicLambda>
const Moose::FunctorBase<T> &
FunctorMaterial::addFunctorProperty(const std::string & name,
                                    PolymorphicLambda my_lammy,
                                    const std::set<ExecFlagType> & clearance_schedule = {
                                        EXEC_ALWAYS});
```

```

> where the first two arguments are used to setup block restriction and the last argument is a lambda
> defining the property evaluation. The lambda must be callable with two arguments, the first

Reviewer 2: ```suggestion
where the first argument will be the functor name stored in the problem functor database (all functor names must be unique), the second argument is a lambda
defining the property evaluation, and the third optional argument is a set defining at which execution stages the functor evaluation cache should be cleared. The lambda must be callable with two arguments, the first
```

corresponding to space, and the second corresponding to time, and must return the type `T` of the
`FunctorMaterialProperty`. An example of setting a constant functor material property that returns

> an `ADReal` looks like:

Reviewer 2: ```suggestion
a `Real` looks like:
```

```c++

>     _constant_unity_prop.setFunctor(

Reviewer 2: ```suggestion
    addFunctorProperty<Real>(
```

>         _mesh, blockIDs(), [](const auto &, const auto &) -> ADReal { return 1.; });

Reviewer 2: ```suggestion
        "foo", [](const auto &, const auto &) -> Real { return 1.; });
```

```

> An example of a functor material property that depends on a nonlinear variable would look like

Reviewer 2: ```suggestion
An example of a functor material property that depends on a fluid properties user object and pressure and temperature functors looks like
```

```c++

>     _u_prop.setFunctor(_mesh, blockIDs(), [this](const auto & r, const auto & t) -> ADReal {
>       return _u_var(r, t);
>     });

Reviewer 2: ```suggestion
  addFunctorProperty<ADReal>(_density_name,
                             [this](const auto & r, const auto & t) -> ADReal
                             { return _fluid.rho_from_p_T(_pressure(r, t), _temperature(r, t)); });
```

```

> In the above example, we simply forward the calling arguments along to the variable. Variable

Reviewer 2: ```suggestion
In the above example, we simply forward the calling arguments along to the captured functors. Variable
```

functor implementation is described in [MooseVariableBase.md#functor-vars]. A test functor material
class to setup a dummy Euler problem is shown in

!listing test/src/materials/ADCoupledVelocityMaterial.C

In the following subsections, we describe the various spatial arguments that functor (material
properties) can be evaluated at. Almost no functor material developers should have to concern
themselves with these details as most material property functions should just appear as functions of
space and time, e.g. the same lambda defining the property evaluation should apply across all
spatial and temporal arguments. However, in the case that a functor material developer wishes to
create specific implementations for specific arguments (as illustrated in `IMakeMyOwnFunctorProps`
test class) or simply wishes to know more about the system, we give the details below.

Any call to a functor (material property) looks like the following
`_foo(const SpatialArg & r, const TemporalArg & t)`. Below are the possible type overloads of
`SpatialArg`.

> ### FaceArg id=spatial-overloads

Reviewer 2: Maybe we should move all of this to the new Functors `index.md`. The following will also need to be updated

Reviewer 2: And I'm happy to help with the updates

Reviewer 2: Oh it looks like we already do have it there. So let's remove this duplicate stuff
